File: service/app/src/main/java/io/miragon/blueprint/adapter/outbound/db/LeasingApplicationPersistenceAdapter.java
package io.miragon.blueprint.adapter.outbound.db;

import io.miragon.blueprint.application.port.outbound.LeasingApplicationRepository;
import io.miragon.blueprint.domain.leasing.ApplicationId;
import io.miragon.blueprint.domain.leasing.LeasingApplication;
import java.util.Optional;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Component;

@Component
public class LeasingApplicationPersistenceAdapter implements LeasingApplicationRepository {

    private final LeasingApplicationJpaRepository repository;

    public LeasingApplicationPersistenceAdapter(LeasingApplicationJpaRepository repository) {
        this.repository = repository;
    }

    @Override
    public LeasingApplication save(LeasingApplication application) {
        LeasingApplicationEntity entity = repository.save(LeasingApplicationEntityMapper.toEntity(application));
        return LeasingApplicationEntityMapper.toDomain(entity);
    }

    @Override
    public Optional<LeasingApplication> findById(ApplicationId id) {
        return repository.findByApplicationId(id.value()).map(LeasingApplicationEntityMapper::toDomain);
    }

    @Override
    public Page findAll(Criteria criteria) {
        // Newest first — the list shows the most recent applications at the top. Spring Data's paging
        // types are used only here, inside the adapter, and never returned to the application layer.
        Pageable pageable = PageRequest.of(criteria.page(), criteria.size(), Sort.by(Sort.Direction.DESC, "createdAt"));
        org.springframework.data.domain.Page<LeasingApplicationEntity> result =
                criteria.status() != null
                        ? repository.findAllByStatus(criteria.status(), pageable)
                        : repository.findAll(pageable);
        return new Page(
                result.getContent().stream().map(LeasingApplicationEntityMapper::toDomain).toList(),
                result.getNumber(),
                result.getSize(),
                result.getTotalElements(),
                result.getTotalPages());
    }
}

File: service/app/src/main/kotlin/io/miragon/blueprint/adapter/outbound/db/LeasingApplicationPersistenceAdapter.kt
package io.miragon.blueprint.adapter.outbound.db

import io.miragon.blueprint.application.port.outbound.LeasingApplicationRepository
import io.miragon.blueprint.domain.leasing.ApplicationId
import io.miragon.blueprint.domain.leasing.LeasingApplication
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Component

@Component
class LeasingApplicationPersistenceAdapter(
    private val repository: LeasingApplicationJpaRepository,
) : LeasingApplicationRepository {

    override fun save(application: LeasingApplication): LeasingApplication {
        val entity = repository.save(LeasingApplicationEntityMapper.toEntity(application))
        return LeasingApplicationEntityMapper.toDomain(entity)
    }

    override fun findById(id: ApplicationId): LeasingApplication? =
        repository.findByApplicationId(id.value)?.let { LeasingApplicationEntityMapper.toDomain(it) }

    override fun findAll(criteria: LeasingApplicationRepository.Criteria): LeasingApplicationRepository.Page {
        // Newest first — the list shows the most recent applications at the top. Spring Data's paging
        // types are used only here, inside the adapter, and never returned to the application layer.
        val pageable = PageRequest.of(criteria.page, criteria.size, Sort.by(Sort.Direction.DESC, "createdAt"))
        val result =
            if (criteria.status != null) {
                repository.findAllByStatus(criteria.status, pageable)
            } else {
                repository.findAll(pageable)
            }
        return LeasingApplicationRepository.Page(
            items = result.content.map { LeasingApplicationEntityMapper.toDomain(it) },
            page = result.number,
            size = result.size,
            totalElements = result.totalElements,
            totalPages = result.totalPages,
        )
    }
}
